### PR TITLE
CORE-1002: `reach update` mach two

### PR DESCRIPTION
**Pull request by mattaudesse**
In addition to CORE-1002, this PR also covers COREs:

- 1104 - `version-compare` should mark 3rd-party images as "connector available" when image is present locally but not tag
- 1103 - Refactor `version-compare2` conditionals
- 1121 - Hook test stubs into hidden `version-compare2` CLI
- 1133 - Document `version-compare --json` + `update-ide`

The bulk of these changes are just JSON stubs for the new `cli-version-compare` test. Its README.md goes into more detail about how these work.
```bash
$ ../scripts/view-log.sh https://100922-205868997-gh.circle-artifacts.com/0/tmp/artifacts/ALGO.cli-version-compare.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   461  100   461    0     0    729      0 --:--:-- --:--:-- --:--:--   729
100   402  100   402    0     0    383      0  0:00:01  0:00:01 --:--:--   383
clean cli-version-compare

clean:
make: Nothing to be done for 'clean'.
build cli-version-compare

build:
make: Nothing to be done for 'build'.
run cli-version-compare

run:
Running unknown-img...                                          ✔ Exited with 1
Running unknown-tag:0.1.7...                                    ✔ Exited with 1
Running unknown-tag:dab71275...                                 ✔ Exited with 1
Running new-images:reach-cli+react-runner+runner...             ✔ Exited with 60
Running new-images:rpc-server#new-tags:reach-cli+devnet-cfx...  ✔ Exited with 60
Running new-tags:reach+react-runner...                          ✔ Exited with 60
Running fully-synchronized...                                   ✔ Exited with 0
Running connector-avail:eth+cfx...                              ✔ Exited with 0
Running connector-avail-3rd-party:image-not-present...          ✔ Exited with 0
Running script#new-images:devnet-algo+postgres...               ✔ Exited with 60
Running script#fully-synchronized...                            ✔ Exited with 60
Running connector-avail-3rd-party:image-present-tag-not...      ✔ Exited with 0
Running connector-avail-3rd-party:image-present-sha-not...      ✔ Exited with 0
```

### CODE REVIEW

File: hs/app/reach/Main.hs [hunk 1]
```diff
@@ -1996,66 +2113,227 @@ versionCompare2 = command "version-compare2" $ info f mempty
...
-                              <$> newDAs <> newTags <> newCAs <> synced
+      let vc@VersionCompare{..} = VersionCompare
+            [VersionCompareIDTs (Image' x) y z | DAQSync x y z <- both]
+            [VersionCompareIDTs (Image' x) y z | DAQNewDigestAvailable x y z <- both]
+            [VersionCompareIDTs (Image' x) y z | DAQNewTags x y z <- both]
+            [VersionCompareIDTs (Image' x) y z | DAQNewConnectorAvailable x y z <- both]
 
-                    let p x = lefty x <> T.replicate (m - T.length (lefty x)) " "
-                    let n a = when (any ((> 0) . length) a) $ putStrLn ""
+      if length uImgs > 0
```
Comment (jeapostrophe) [inline hs/app/reach/Main.hs:2126]: Eww, I really hate the way `if then else` looks in Haskell. The formatting always messes me up to the point that I always use `case`. I also often do `case (x, y, z)` so I can get them all at the same indentation too

**Review by jeapostrophe — APPROVED** [review]

File: hs/app/reach/Main.hs (hunk 1 above)
Comment (mattaudesse) [reply hs/app/reach/Main.hs:2126]: I really don't like how this looks either but `hs-format` thinks it's better somehow. I thought this was way easier to read:

https://github.com/reach-sh/reach-lang/commit/2518e01b7c0094340d00bbcc2ff22973dd80afd9#diff-ceefed8a152f4a53d8a11bd2bc7dabb608c98d48cce9c3d0f4484a0451b6ab26L1838-L1881

```haskell
 liftIO $ do
    if length uImgs > 0 then do
      forM_ uImgs $ \a -> T.putStrLn $ "Unknown image: " <> lefty a <> "."
      T.putStrLn $ "Please open an issue at " <> uriIssues <> " including the failures listed above."
      exitWith $ ExitFailure 1

    else if length uTags > 0 then do
      forM_ uTags $ \(x, y) -> T.putStrLn $ "Unknown tag: " <> y <> " for image: " <> lefty x <> "."
      exitWith $ ExitFailure 1

    else if length (synced <> newCAs) >= length both then do
      putStrLn "Reach is up-to-date."
      exitWith $ ExitFailure 0

    else do
      ...
```

I'll try to tidy it up with a `case` but I think that might look really crowded for this block of code.

File: examples/cli-version-compare/run.sh [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+e 60 new-images:reach-cli+react-runner+runner
+e 60 new-images:rpc-server#new-tags:reach-cli+devnet-cfx
+e 60 new-tags:reach+react-runner
+e  0 fully-synchronized
+e  0 connector-avail:eth+cfx
+e  0 connector-avail-3rd-party:image-not-present
+e 60 script#new-images:devnet-algo+postgres
+e 60 script#fully-synchronized
+e  0 connector-avail-3rd-party:image-present-tag-not
+e  0 connector-avail-3rd-party:image-present-sha-not
```
Comment (jeapostrophe) [inline examples/cli-version-compare/run.sh:60]: I would prefer something that looks at the every subdirectory and gets the expected error code from something in the directory. I worry about someone thinking they added a test case (by putting in a subdir) but forgetting to add it here

File: examples/cli-version-compare/script#fully-synchronized/o.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Reach's Docker images are up-to-date.
```
Comment (jeapostrophe) [inline examples/cli-version-compare/script#fully-synchronized/o.txt:1]: Did you mean to commit the empty [examples/cli-version-compare/script#fully-synchronized/reach](https://github.com/reach-sh/reach-lang/pull/557/files#diff-3bc127310e6b35ed9c9d3c87fdeb49b4cf31e70d20ba4eb615b5ab237d908ab3)?

File: examples/cli-version-compare/script#new-images:devnet-algo+postgres/o.txt [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+The following images are fully synchronized:
+ * reachsh/reach         fd08459b:  stable, 0, 0.1, 0.1.8, 34107655, 2022-01-31
+ * reachsh/reach-cli     bf5922b9:  latest
+ * reachsh/react-runner  1c35abc8:  stable, 0, 0.1, 0.1.8, 34107655, 2022-01-31
+ * reachsh/rpc-server    f70fd052:  stable, 0, 0.1, 0.1.8, 34107655, 2022-01-31
+ * reachsh/runner        90bbb280:  stable, 0, 0.1, 0.1.8, 34107655, 2022-01-31
+ * reachsh/devnet-cfx    e9f029f0:  stable, 0, 0.1, 0.1.8, 34107655, 2022-01-31
+ * reachsh/devnet-eth    110bc4d6:  stable, 0, 0.1, 0.1.8, 34107655, 2022-01-31
+
+There's a newer version of the `reach` script available.
```
Comment (jeapostrophe) [inline examples/cli-version-compare/script#new-images:devnet-algo+postgres/o.txt:14]: Another reach file here...?

**Review by jeapostrophe — APPROVED** [review]
There are a few comments. Not necessarily needed before merging, but I won't press the button.

I think I understand correctly that neither of these commands are exposed to users yet on purpose (no docs and not connected to `reach update`/`reach upgrade`) but that is the next step?

**Comment by mattaudesse** [comment]
> I think I understand correctly that neither of these commands are exposed to users yet on purpose (no docs and not connected to `reach update`/`reach upgrade`) but that is the next step?

Yep, these are big enough I figured we'd want to regroup and assess before I try calling them `reach info` or a replacement `reach update`/`reach upgrade`.

File: examples/cli-version-compare/run.sh (hunk 2 above)
Comment (mattaudesse) [reply examples/cli-version-compare/run.sh:60]: I found this more convenient when focusing on a single test and less scattered but I'll make a separate PR if you think that's better.

File: examples/cli-version-compare/script#fully-synchronized/o.txt (hunk 3 above)
Comment (mattaudesse) [reply examples/cli-version-compare/script#fully-synchronized/o.txt:1]: These are intentional, yeah - they pretend there's a newer version of the script available. `README.md` and `run.sh` explain how/why.

File: examples/cli-version-compare/run.sh (hunk 2 above)
Comment (mattaudesse) [reply examples/cli-version-compare/run.sh:60]: I've created CORE-1134 to track it.

**Pull request merged**